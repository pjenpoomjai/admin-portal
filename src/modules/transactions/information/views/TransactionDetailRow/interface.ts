import { ResourceKey } from 'contexts/resource/interface'

export interface ITransactionDetailRow<T> {
	id: string
	label: string
	keys: (keyof T)[]
	value: T
	customText?: () => string
}

export interface ITransactionDetailRowProps<T> {
	data: ITransactionDetailRow<T>
	resourceValueKey?: ResourceKey
}

export interface ITransactionDetailRowsProps<T> {
	id: string
	data: ITransactionDetailRow<T>[]
	resourceValueKey?: ResourceKey
}
