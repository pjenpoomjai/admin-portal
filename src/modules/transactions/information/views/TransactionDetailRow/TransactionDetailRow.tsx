import { Typography } from 'components/common'
import {
	ITransactionDetailRowProps,
	ITransactionDetailRowsProps,
} from './interface'
import { Fragment } from 'react'
import { renderText } from '../TransactionInformation/TransactionInformation'

const TransactionDetailRow = <T,>(props: ITransactionDetailRowProps<T>) => {
	const { data, resourceValueKey } = props
	const { id, label, keys, value, customText } = data

	if (keys.length === 1) {
		return (
			<div className="flex w-full">
				<div className="flex w-1/2 py-2 px-6">
					<Typography id={id} className="font-semibold break-words">
						{label}
					</Typography>
				</div>
				<div className="flex w-1/2 p-2">
					{renderText({
						id: `${id}-${String(keys[0])}`,
						value: customText ? customText() : (value[keys[0]] as string),
						resourceValueKey,
					})}
				</div>
			</div>
		)
	}

	const widthValue = keys.length === 2 ? 'w-2/5' : 'w-[calc((4/15)*100%)]'

	return (
		<div className="flex w-full">
			<div className="flex w-1/5 py-2 px-6">
				<Typography id={id} className="font-semibold break-words">
					{label}
				</Typography>
			</div>
			{keys.map((k) => {
				const displayValue = customText ? customText() : (value[k] as string)
				return (
					<div className={`flex ${widthValue} p-2`} key={`${id}-${String(k)}`}>
						{renderText({
							id: `${id}-${String(k)}`,
							value: displayValue,
							resourceValueKey,
						})}
					</div>
				)
			})}
		</div>
	)
}

export const TransactionDetailRows = <T,>(
	props: ITransactionDetailRowsProps<T>,
) => {
	const { data, id, resourceValueKey } = props

	return data.map((d) => {
		return (
			<Fragment key={`${id}-${d.id}`}>
				{TransactionDetailRow({ data: d, resourceValueKey })}
			</Fragment>
		)
	})
}

export default TransactionDetailRow
