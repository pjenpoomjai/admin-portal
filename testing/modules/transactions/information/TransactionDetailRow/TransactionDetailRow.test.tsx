import { render } from '@testing-library/react'
import { TransactionDetailRows } from 'modules/transactions/information/views/TransactionDetailRow/TransactionDetailRow'

describe('TransactionDetailRow : View', () => {
	test('Should render successfully 1 key', () => {
		const { baseElement } = render(
			TransactionDetailRows({
				id: 'row-1',
				data: [
					{
						id: 'r1',
						keys: ['key-1'],
						label: '',
						value: {
							'key-1': 'value-1',
						},
					},
					{
						id: 'r2',
						keys: ['key-1'],
						label: '',
						value: {
							'key-1': 'value-1',
						},
						customText: () => {
							return 'custom'
						},
					},
				],
			}),
		)
		expect(baseElement).toMatchSnapshot()
	})

	test('Should render successfully 2 keys', () => {
		const { baseElement } = render(
			TransactionDetailRows({
				id: 'row-1',
				data: [
					{
						id: 'r1',
						keys: ['key-1', 'key-2'],
						label: '',
						value: {
							'key-1': 'value-1',
							'key-2': 'value-2',
						},
					},
					{
						id: 'r2',
						keys: ['key-1', 'key-2'],
						label: '',
						value: {
							'key-1': 'value-1',
							'key-2': 'value-2',
						},
						customText: () => {
							return 'custom'
						},
					},
				],
			}),
		)
		expect(baseElement).toMatchSnapshot()
	})

	test('Should render successfully 3 keys', () => {
		const { baseElement } = render(
			TransactionDetailRows({
				id: 'row-1',
				data: [
					{
						id: 'r1',
						keys: ['key-1', 'key-2', 'key-3'],
						label: '',
						value: {
							'key-1': 'value-1',
							'key-2': 'value-2',
							'key-3': 'value-3',
						},
					},
					{
						id: 'r2',
						keys: ['key-1', 'key-2', 'key-3'],
						label: '',
						value: {
							'key-1': 'value-1',
							'key-2': 'value-2',
							'key-3': 'value-3',
						},
						customText: () => {
							return 'custom'
						},
					},
				],
			}),
		)
		expect(baseElement).toMatchSnapshot()
	})
})
