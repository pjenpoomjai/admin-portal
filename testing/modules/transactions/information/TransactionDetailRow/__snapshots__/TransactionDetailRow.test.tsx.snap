// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionDetailRow : View Should render successfully 1 key 1`] = `
<body>
  <div>
    <div
      class="flex w-full"
    >
      <div
        class="flex w-1/2 py-2 px-6"
      >
        <p
          class="block antialiased font-sans text-base leading-relaxed text-inherit font-semibold break-words"
          data-id="r1-typography"
          data-testid="r1-typography"
          placeholder=""
        />
      </div>
      <div
        class="flex w-1/2 p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r1-key-1-typography"
          data-testid="r1-key-1-typography"
          notmatchlabel="-"
          placeholder=""
        >
          value-1
        </p>
      </div>
    </div>
    <div
      class="flex w-full"
    >
      <div
        class="flex w-1/2 py-2 px-6"
      >
        <p
          class="block antialiased font-sans text-base leading-relaxed text-inherit font-semibold break-words"
          data-id="r2-typography"
          data-testid="r2-typography"
          placeholder=""
        />
      </div>
      <div
        class="flex w-1/2 p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r2-key-1-typography"
          data-testid="r2-key-1-typography"
          notmatchlabel="-"
          placeholder=""
        >
          custom
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`TransactionDetailRow : View Should render successfully 2 keys 1`] = `
<body>
  <div>
    <div
      class="flex w-full"
    >
      <div
        class="flex w-1/5 py-2 px-6"
      >
        <p
          class="block antialiased font-sans text-base leading-relaxed text-inherit font-semibold break-words"
          data-id="r1-typography"
          data-testid="r1-typography"
          placeholder=""
        />
      </div>
      <div
        class="flex w-2/5 p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r1-key-1-typography"
          data-testid="r1-key-1-typography"
          notmatchlabel="-"
          placeholder=""
        >
          value-1
        </p>
      </div>
      <div
        class="flex w-2/5 p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r1-key-2-typography"
          data-testid="r1-key-2-typography"
          notmatchlabel="-"
          placeholder=""
        >
          value-2
        </p>
      </div>
    </div>
    <div
      class="flex w-full"
    >
      <div
        class="flex w-1/5 py-2 px-6"
      >
        <p
          class="block antialiased font-sans text-base leading-relaxed text-inherit font-semibold break-words"
          data-id="r2-typography"
          data-testid="r2-typography"
          placeholder=""
        />
      </div>
      <div
        class="flex w-2/5 p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r2-key-1-typography"
          data-testid="r2-key-1-typography"
          notmatchlabel="-"
          placeholder=""
        >
          custom
        </p>
      </div>
      <div
        class="flex w-2/5 p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r2-key-2-typography"
          data-testid="r2-key-2-typography"
          notmatchlabel="-"
          placeholder=""
        >
          custom
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`TransactionDetailRow : View Should render successfully 3 keys 1`] = `
<body>
  <div>
    <div
      class="flex w-full"
    >
      <div
        class="flex w-1/5 py-2 px-6"
      >
        <p
          class="block antialiased font-sans text-base leading-relaxed text-inherit font-semibold break-words"
          data-id="r1-typography"
          data-testid="r1-typography"
          placeholder=""
        />
      </div>
      <div
        class="flex w-[calc((4/15)*100%)] p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r1-key-1-typography"
          data-testid="r1-key-1-typography"
          notmatchlabel="-"
          placeholder=""
        >
          value-1
        </p>
      </div>
      <div
        class="flex w-[calc((4/15)*100%)] p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r1-key-2-typography"
          data-testid="r1-key-2-typography"
          notmatchlabel="-"
          placeholder=""
        >
          value-2
        </p>
      </div>
      <div
        class="flex w-[calc((4/15)*100%)] p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r1-key-3-typography"
          data-testid="r1-key-3-typography"
          notmatchlabel="-"
          placeholder=""
        >
          value-3
        </p>
      </div>
    </div>
    <div
      class="flex w-full"
    >
      <div
        class="flex w-1/5 py-2 px-6"
      >
        <p
          class="block antialiased font-sans text-base leading-relaxed text-inherit font-semibold break-words"
          data-id="r2-typography"
          data-testid="r2-typography"
          placeholder=""
        />
      </div>
      <div
        class="flex w-[calc((4/15)*100%)] p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r2-key-1-typography"
          data-testid="r2-key-1-typography"
          notmatchlabel="-"
          placeholder=""
        >
          custom
        </p>
      </div>
      <div
        class="flex w-[calc((4/15)*100%)] p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r2-key-2-typography"
          data-testid="r2-key-2-typography"
          notmatchlabel="-"
          placeholder=""
        >
          custom
        </p>
      </div>
      <div
        class="flex w-[calc((4/15)*100%)] p-2"
      >
        <p
          class="block antialiased font-sans text-base font-light leading-relaxed text-inherit break-words"
          data-id="r2-key-3-typography"
          data-testid="r2-key-3-typography"
          notmatchlabel="-"
          placeholder=""
        >
          custom
        </p>
      </div>
    </div>
  </div>
</body>
`;
